### PR TITLE
SelectParser: add support for grouping labels

### PR DESCRIPTION
**Pull request by radu-gheorghe**
The Select (JSON) syntax didn't support a way to change the label of grouping results. [The YQL variant does](https://docs.vespa.ai/en/reference/querying/grouping-language.html#labels) with something like:

```
each(...) as(mylabel)
```

This commit fixes this gap. Something like this now works:

```json
{
  "all": {
    "group": "explanation_attribute",
    "each": {
      "output": "count()",
      "label": "explanations"
    }
  }
}
```

Note how the label is within `each`. Feel free to disagree with that :) It's in the name of simplicity/reliability, because JSON grouping is translated to "normal" grouping, where `as` needs to go **after** `each` (IIUC). The current implementation guarantees that `each` is there. If we did it one level higher (which is more natural), we'd have to do more validation to make sure we don't put `as` in the wrong place.

Speaking of validation, we do a reasonable job (e.g., reject the request if `label` is upper in the hierarchy - see tests for more), but we don't cover everything. AFAIK, the only corner case is when someone has an array for `each` (which is unlikely, IMO):
```json
    "each": [{
      "output": "count()",
      "label": "explanations"
    }]
```

In this case, we silently drop the label instead of rejecting the request as invalid. I think this is in line with the rather loose validation that we do for JSON grouping in general: we rely on the underlying grouping syntax parser to catch some of the errors. You can argue that's not nice, but refactoring for better JSON validation is a separate task, IMO.

### CODE REVIEW

**Review by boeker — DISMISSED** [review]
Seems reasonable to me 🙂

**Comment by johansolbakken** [comment]
Looks good to me 👍🏻 , but should we run it by the architecture meeting as well?

**Comment by radu-gheorghe** [comment]
Ah, that would be a good idea 🙈 I'll do it later.

**Review by bjorncs — DISMISSED** [review]

**Comment by radu-gheorghe** [comment]
OK, I've pushed the second version, which aligns better with what we discussed.

Namely, `label` now goes one level up. Like:

```JSON
    "grouping": [
      {
        "all": {
          "group": "year",
          "label": "by_year",
          "each": {
            "output": "count()",
          }
        }
      }
    ]
```

We shouldn't be missing validation cases now, either.

@boeker @johansolbakken - what do you think?

**Comment by boeker** [comment]
In its current form, this breaks the `RequestBuilderTestCase.requireThatExpressionsCanBeLabeled` test case. This is caused by the change in `RequestBuilder`. What exactly does that change do? 🙂 

And I am also wondering why the `GroupingParser.jj` has to be changed. It probably has something to do with the translation from Select to Grouping. What makes that necessary? 🤔

**Comment by radu-gheorghe** [comment]
Uh, I'm so sorry for missing the basic stuff (e.g., failed test).

The parser&builder changes were an artifact of my initial understanding, that we need to change the grammar in order to be able to label grouping statements with no `each`, like cardinality requests. But it looks like they are already supported, for example:

```
all(
  group(genres_att)
  output( count() as(genres_total) )
)
```

Which is naturally supported in Select (no change needed):
```JSON
"grouping": [
  {
    "all": {
      "group": "genres_att",
      "output": "count() as(genres_total)"
    }
  }
]
```

So I've reverted the grammar change. Now the PR should add the ability to SelectParser to parse a request like:
```JSON
"grouping": [
  {
    "all": {
      "group": "year",
      "label": "by_year",
      "each": {
        "output": "count()"
      }
    }
  }
]
```

Into something like:
```
all(
  group(year)
  each(output(count()))
  as(by_year)
)
```

Hopefully, I haven't missed other things as well 🙈

File: container-search/src/main/java/com/yahoo/search/query/SelectParser.java [hunk 1]
```diff
@@ -358,21 +359,53 @@ private String toGroupingRequest(Inspector groupingJson) {
...
                 groupingJson.traverse((ArrayTraverser) (index, item) -> {
-                    toGroupingRequest(item, b);
+                    toGroupingRequest(item, parentKey, false, b);
                     if (index + 1 < groupingJson.entries())
                         b.append(",");
                 });
                 break;
             case OBJECT:
+                String label = isOperationBody ? readLabel(groupingJson) : null;
+                boolean[] directEachSeen = {false};
```
Comment (boeker) [inline container-search/src/main/java/com/yahoo/search/query/SelectParser.java:380]: Using an array here is a bit ugly, but I don't think there is a much better solution. 😅

**Review by boeker — APPROVED** [review]
Looks good to me. :)

File: container-search/src/main/java/com/yahoo/search/query/SelectParser.java (hunk 1 above)
Comment (radu-gheorghe) [reply container-search/src/main/java/com/yahoo/search/query/SelectParser.java:380]: I agree with both sides of the argument :)

**Review by johansolbakken — APPROVED** [review]
Me too 👍🏻
